src/ripple/shamap/impl/SHAMapTreeNode.cpp

```diff
@@ -593,5 +712,190 @@ SHAMapInnerNode::canonicalizeChild(int branch, std::shared_ptr<SHAMapAbstractNod
     return node;
 }
 
+bool
+SHAMapInnerNodeV2::has_common_prefix(uint256 const& key) const
+{
+    auto x = common_.begin();
+    auto y = key.begin();
+    for (unsigned i = 0; i < depth_/2; ++i, ++x, ++y)
+    {
+        if (*x != *y)
+            return false;
+    }
+    if (depth_ & 1)
+    {
+        auto i = depth_/2;
+        return (*(common_.begin() + i) & 0xF0) == (*(key.begin() + i) & 0xF0);
+    }
+    return true;
+}
+
+int
+SHAMapInnerNodeV2::get_common_prefix(uint256 const& key) const
+{
+    auto x = common_.begin();
+    auto y = key.begin();
+    auto r = 0;
+    for (unsigned i = 0; i < depth_/2; ++i, ++x, ++y, r += 2)
+    {
+        if (*x != *y)
+        {
+            if ((*x & 0xF0) == (*y & 0xF0))
+                ++r;
+            return r;
+        }
+    }
+    if (depth_ & 1)
+    {
+        auto i = depth_/2;
+        if ((*(common_.begin() + i) & 0xF0) == (*(key.begin() + i) & 0xF0))
+            ++r;
+    }
+    return r;
+}
+
+void
+SHAMapInnerNodeV2::setChildren(std::shared_ptr<SHAMapTreeNode> const& child1,
+                               std::shared_ptr<SHAMapTreeNode> const& child2)
+{
+    assert(child1->peekItem()->key() != child2->peekItem()->key());
+    auto k1 = child1->peekItem()->key().begin();
+    auto k2 = child2->peekItem()->key().begin();
+    auto k = common_.begin();
+    for (depth_ = 0; *k1 == *k2; ++depth_, ++k1, ++k2, ++k)
+        *k = *k1;
+    unsigned b1;
+    unsigned b2;
+    if ((*k1 & 0xF0) == (*k2 & 0xF0))
+    {
+        *k = *k1 & 0xF0;
+        b1 = *k1 & 0x0F;
+        b2 = *k2 & 0x0F;
+        depth_ = 2*depth_ + 1;
+    }
+    else
+    {
+        b1 = *k1 >> 4;
+        b2 = *k2 >> 4;
+        depth_ = 2*depth_;
+    }
+    mChildren[b1] = child1;
+    mIsBranch |= 1 << b1;
+    mChildren[b2] = child2;
+    mIsBranch |= 1 << b2;
+}
+
+void
+SHAMapInnerNodeV2::set_common(int depth, uint256 const& common)
+{
+    depth_ = depth;
+    common_ = common;
+}
+
+uint256 const&
+SHAMapInnerNode::key() const
+{
+    Throw<std::logic_error>("SHAMapInnerNode::key() should never be called");
+    static uint256 x;
+    return x;
+}
+
+uint256 const&
+SHAMapInnerNodeV2::key() const
+{
+    return common_;
+}
+
+uint256 const&
+SHAMapTreeNode::key() const
+{
+    return mItem->key();
+}
+
+void
+SHAMapInnerNode::invariants(bool is_v2, bool is_root) const
+{
+    assert(!is_v2);
+    assert(mType == tnINNER);
+    unsigned count = 0;
+    for (int i = 0; i < 16; ++i)
+    {
+        if (mHashes[i].isNonZero())
+        {
+            assert((mIsBranch & (1 << i)) != 0);
+            if (mChildren[i] != nullptr)
+                mChildren[i]->invariants(is_v2);
+            ++count;
+        }
+        else
+        {
+            assert((mIsBranch & (1 << i)) == 0);
+        }
+    }
+    if (!is_root)
+    {
+        assert(mHash.isNonZero());
+        assert(count >= 1);
+    }
+    if (count == 0)
+        assert(mHash.isZero());
+    else
+        assert(mHash.isNonZero());
```

Comment:
This is going to compile weirdly in fully optimized builds:
```
if (count == 0)
{
}
else
{
}
```

Consider this instead
```
assert((count == 0) ? mHash.isZero() : mHash.isNonZero());
```